### PR TITLE
MGMT-7806: Change host events API in order to solve segmentation fault

### PR DESCRIPTION
**Pull request by yevgeny-shnaidman**
Currently host/cluster events store ClusterID as strfmt.UUID. As a result, APIs
that create events also expect clusterId parameter as strfmt.UUID. In case of
unbound host this may cause issues, since ClusterID will be nil.
We change the host event to store ClusterID as pointer, and propagate it as is
to SendHostEvent* APIs, which already handle the case where it is nil

# Assisted Pull Request

## List all the issues related to this PR

- [ ] New Feature
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Assignees

/cc @
/cc @

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] Reviewers have been listed
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- [ ] Are the title and description (in both PR and commit) meaningful and clear?
- [ ] Is there a bug required (and linked) for this change?
- [ ] Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: tools/generate_events.py [hunk 1]
```diff
@@ -96,7 +96,7 @@
 }
 {%- else %}
 func (e *{{eventName}}) GetClusterId() *strfmt.UUID {
-    {% if event.properties.cluster_id -%}     return &e.ClusterId
+    {% if event.properties.cluster_id -%}     return e.ClusterId
```
Comment (masayag) [inline tools/generate_events.py:99]: Could you describe the issues that occurred in the unbound host ? 
If it reduced only to invocation of this method, perhaps it can be changed to
```
    {% if event.properties.cluster_id -%}     if len(e.ClusterId) > 0 {
        return &e.ClusterId
    } else {
      return nil
    }
```
to be nil-safe.

If the problem with with the call to the eventshandler.SendHostEvent(...), then there is a need to change their signatures to support nil values as cluster-id.

**Comment by masayag** [comment]
pls fix typo (Chahe) in PR message and in the commit message as well.

File: internal/host/host.go [hunk 2]
```diff
@@ -830,10 +830,10 @@ func (m *Manager) ResetHost(ctx context.Context, h *models.Host, reason string,
 	defer func() {
 		if shouldAddEvent {
 			if isFailed {
-				eventgen.SendHostResetInstallationFailedEvent(ctx, m.eventsHandler, h.InfraEnvID, *h.ID,
+				eventgen.SendHostResetInstallationFailedEvent(ctx, m.eventsHandler, h.ClusterID, *h.ID,
```
Comment (nmagnezi) [inline internal/host/host.go:833]: good catch

File: tools/generate_events.py (hunk 1 above)
Comment (yevgeny-shnaidman) [reply tools/generate_events.py:99]: The specific issue was with the call to the SendHostValidationFallingEvent function (but it may happen to other calls for host events). in previous implementation SendHostValidationFallingEvent received clusterId as strfmt.UUID. in case of the unbound host, clusterId is nil, so calling SendHostValidationFallingEvent (*host.ClusterID) caused segmentation fault. I think it is better simply to store ClusterID as pointer in the host events, since we already have a code that checks if it is nil and calls AddEvent function with appropriate params

Comment (masayag) [reply tools/generate_events.py:99]: Okay, in that case @nmagnezi we'll have to use the practice for the infra-env event type which may have the cluster-id optionals in that context

**Review by nmagnezi — CHANGES_REQUESTED** [review]
looks good overall, but this fixes an issue that we didn't have tests for.
please add unit-test(s) to trigger those events with nil cluster ID.

**Comment by yevgeny-shnaidman** [comment]
> looks good overall, but this fixes an issue that we didn't have tests for.
> please add unit-test(s) to trigger those events with nil cluster ID.

added

**Comment by masayag** [comment]
LGTM, @nmagnezi pls take another look.

**Review by nmagnezi — APPROVED** [review]
/lgtm
test looks good. thanks!
